TypeScript.py

```diff
@@ -1501,6 +1531,92 @@ def run(self, text):
+    # set to make sure the state reset to be done before run
+    reset_already = False
+    input_text = ""
+
+    @classmethod
+    def reset(cls):
+        cls.navto_panel_started = False
+        cls.insert_text_finished = False
+
+    def is_enabled(self):
+        return is_typescript(self.window.active_view())
+
+    def run(self, input_text = ""):
+        logger.log.debug("start running navto wiht text: %s" % input_text)
+
+        NavToCommand.reset_already = False
+        
+        self.window.run_command("hide_overlay")
+
+        if NavToCommand.reset_already:
+            NavToCommand.reset_already = False
+        else:
+            logger.log.debug("reset in run")
+            NavToCommand.reset()
+            NavToCommand.reset_already = True
+
+        NavToCommand.input_text = input_text        
+        NavToCommand.navto_panel_started = True
+
+        # query text is not always equal to input text. When no input_text is given
+        # alternative query text is used to ensure the panel stay active
+        query_text = "a" if input_text == "" else input_text
+        items = cli.service.navTo(query_text, self.window.active_view().file_name())
+        self.items = items if len(items) != 0 else self.items
+
+        self.window.show_quick_panel(self.format_navto_result(self.items), self.on_done)  
+        logger.log.debug("end running navto wiht text: %s" % input_text)
+        
+    def on_done(self, index):
+        logger.log.debug("enter on_done. input_text: %s, started: %s, insert_text_finished: %s" % 
+               (NavToCommand.input_text, NavToCommand.navto_panel_started, NavToCommand.insert_text_finished))
+        
+        if NavToCommand.reset_already:
+            NavToCommand.reset_already = False
+        else:
+            logger.log.debug("reset in on_done")
+            NavToCommand.reset()
+            NavToCommand.reset_already = True
+
+        if index >= 0:
+            item = self.items[index]
+            line, offset = item['start']['line'], item['start']['offset']
+            file_at_location = item['file'] + ":%s:%s" % (line, offset)
+            self.window.open_file(file_at_location, sublime.ENCODED_POSITION)
+
+        logger.log.debug("exit on_done. input_text: %s, started: %s, insert_text_finished: %s" % 
+               (NavToCommand.input_text, NavToCommand.navto_panel_started, NavToCommand.insert_text_finished))
+
+    def format_navto_result(self, item_list):
+        return [
+                [i['name'], 
+                "%s in %s %s" % (
+                    i['kind'],
+                    (i["containerKind"] if "containerKind" in i else os.path.basename(i["file"]) + " (global)"), 
+                    (i["containerName"] if "containerName" in i else "")
+                )] 
+                for i in item_list]
+
+    def on_highlight(self, index):
+        pass
+
+class InsertTextCommand(sublime_plugin.TextCommand):
```

Comment:
Do we need this?  Why not use one of the existing Sublime commands to insert/append text?